src/blocks/service_status.rs

```diff
@@ -0,0 +1,173 @@
+//! `active_state` | A valid [`State`] | [`State::Idle`]
+//! `inactive_state` | A valid [`State`]  | [`State::Critical`]
+//!
+//! Placeholder    | Value                     | Type   | Unit
+//! ---------------|---------------------------|--------|-----
+//! `service`      | The name of the service   | Text   | -
+//!
+//! # Example
+//!
+//! Example using an icon:
+//!
+//! ```toml
+//! [[block]]
+//! block = "service_status"
+//! service = "cups"
+//! active_format = " ^icon_tea "
+//! inactive_format = " no ^icon_tea "
+//! ```
+//!
+//! Example overriding the default `inactive_state`:
+//!
+//! ```toml
+//! [[block]]
+//! block = "service_status"
+//! service = "shadow"
+//! active_format = ""
+//! inactive_format = " Integrity of password and group files failed "
+//! inactive_state = "Warning"
+//! ```
+//!
+
+use super::prelude::*;
+use zbus::dbus_proxy;
+use zbus::PropertyStream;
+
+#[derive(Deserialize, Debug, Default)]
+#[serde(default)]
+pub struct Config {
+    driver: DriverType,
+    service: String,
+    active_format: FormatConfig,
+    inactive_format: FormatConfig,
+    active_state: Option<State>,
+    inactive_state: Option<State>,
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(rename_all = "snake_case")]
+enum DriverType {
+    #[default]
+    Systemd,
+}
+
+pub async fn run(config: Config, mut api: CommonApi) -> Result<()> {
+    api.event_receiver.close();
+
+    let mut widget = Widget::new();
+    let active_format = config.active_format.with_default(" $service active ")?;
+    let inactive_format = config.inactive_format.with_default(" $service inactive ")?;
+
+    let active_state = config.active_state.unwrap_or(State::Idle);
+    let inactive_state = config.inactive_state.unwrap_or(State::Critical);
+
+    let mut driver: Box<dyn Driver> = match config.driver {
+        DriverType::Systemd => Box::new(SystemdDriver::new(config.service.clone()).await?),
+    };
+
+    loop {
+        let service_active_state = driver.is_active().await?;
+
+        if service_active_state {
+            widget.state = active_state;
+            widget.set_format(active_format.clone());
+        } else {
+            widget.state = inactive_state;
+            widget.set_format(inactive_format.clone());
+        };
+
+        widget.set_values(map!(
+            "service" =>Value::text(config.service.clone()),
+        ));
+        api.set_widget(&widget).await?;
+
+        driver.wait_for_change().await?;
+    }
+}
+
+#[async_trait]
+trait Driver {
+    async fn is_active(&self) -> Result<bool>;
+    async fn wait_for_change(&mut self) -> Result<()>;
+}
+
+struct SystemdDriver {
+    proxy: UnitProxy<'static>,
+    active_state_changed: PropertyStream<'static, String>,
+}
+
+impl SystemdDriver {
+    async fn new(service: String) -> Result<Self> {
+        let dbus_conn = new_system_dbus_connection().await?;
```

Comment:
can we support user services too?

Reply:
I did try that, but user services are removed from dbus (the path is gone) when the service is stopped. I couldn't find a nice way to discover that it had come back. Perhaps if we did a filter like we do in the music block... but it feels like overkill if you are only interested in 1 service.

Reply:
Interesting. Guess if it is going to require a flag in the config then can be revisited in the future without having to figure it out now.